sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfigurationprovider/_azureappconfigurationprovider.py

```
@@ -0,0 +1,180 @@
+                connection_string, user_agent=useragent, headers=headers)
+            return
+        self._client = AzureAppConfigurationClient(
+            endpoint, credential, user_agent=useragent, headers=headers)
+
+    def __resolve_keyvault_references(self, config, key_vault_options, secret_clients):
+        if key_vault_options is None:
+            warnings.warn(
+                "Key Vault Reference found, but no Key Vault Options were provided")
+            return
+        j_object = json.loads(config.value)
+        uri_value = j_object['uri']
+
+        uri = urlparse(uri_value)
+
+        key_vault_uri = "https://" + uri.hostname
+        key_vault_secret_prefix = "/secrets/"
+        key_vault_secret_name = uri.path[len(
+            key_vault_secret_prefix):]
+        if key_vault_options.credential is not None:
+            secret_client = None
+
+            # Clients only should be made once, will reuse if client already made
+            for client_uri in secret_clients:
+                if client_uri == key_vault_uri:
+                    secret_client = secret_clients[client_uri]
+                    break
+            if secret_client is None:
+                secret_client = SecretClient(
+                    vault_url=key_vault_uri, credential=key_vault_options.credential)
+                secret_clients[key_vault_uri] = secret_client
+            secret = secret_client.get_secret(
+                key_vault_secret_name)
+            self._dict[self.__trim(config.key)] = secret.value
+            return
+        if key_vault_options.secret_clients is not None:
+            for secret_client in key_vault_options.secret_clients:
+                if secret_client._vault_url == key_vault_uri:
+                    secret = secret_client.get_secret(
+                        key_vault_secret_name)
+                    self._dict[self.__trim(
+                        config.key)] = secret.value
+                break
+            return
```

Jsut curious, is it possible that a secret client URI will not match any of the `vault_uri`s  for `keyvault_options.secret_clients` ? If that's the case and the user supplies a secret resolver too, wouldn't we want to try that?

---

I think this goes back to a comment during the demo. It is possible. It should throw an error and stop all processing if we can't resolve. We may want to add a fail-fast option to stop that though.